src/coreclr/System.Private.CoreLib/src/Internal/Runtime/Binder/AssemblyBinderCommon.cs

```diff
@@ -0,0 +1,862 @@
+                    return PEKind.AMD64;
+                else
+                {
+                    // We don't support other architectures
+                    throw new BadImageFormatException();
+                }
+            }
+            else
+            {
+                // 32-bit, non-agnostic
+                if (dwImageType == IMAGE_FILE_MACHINE_I386)
+                    return PEKind.I386;
+                else if (dwImageType == IMAGE_FILE_MACHINE_ARMNT)
+                    return PEKind.ARM;
+                else
+                {
+                    // Not supported
+                    throw new BadImageFormatException();
+                }
+            }
+        }
+
+        // HResult
+        private const int RO_E_METADATA_NAME_NOT_FOUND = unchecked((int)0x8000000F);
+        private const int E_PATHNOTFOUND = unchecked((int)0x80070003);
+        private const int E_NOTREADY = unchecked((int)0x80070015);
+        private const int E_BADNETPATH = unchecked((int)0x80070035);
+        private const int E_BADNETNAME = unchecked((int)0x80070043);
+        private const int E_INVALID_NAME = unchecked((int)0x8007007B);
+        private const int E_MODNOTFOUND = unchecked((int)0x8007007E);
+        private const int E_DLLNOTFOUND = unchecked((int)0x80070485);
+        private const int E_WRONG_TARGET_NAME = unchecked((int)0x80070574);
+        private const int INET_E_CANNOT_CONNECT = unchecked((int)0x800C0004);
+        private const int INET_E_RESOURCE_NOT_FOUND = unchecked((int)0x800C0005);
+        private const int INET_E_OBJECT_NOT_FOUND = unchecked((int)0x800C0006);
+        private const int INET_E_DATA_NOT_AVAILABLE = unchecked((int)0x800C0007);
+        private const int INET_E_DOWNLOAD_FAILURE = unchecked((int)0x800C0008);
+        private const int INET_E_CONNECTION_TIMEOUT = unchecked((int)0x800C000B);
+        private const int INET_E_UNKNOWN_PROTOCOL = unchecked((int)0x800C000D);
+        private const int FUSION_E_APP_DOMAIN_LOCKED = unchecked((int)0x80131053);
+        private const int CLR_E_BIND_ASSEMBLY_VERSION_TOO_LOW = unchecked((int)0x80132000);
+        private const int CLR_E_BIND_ASSEMBLY_PUBLIC_KEY_MISMATCH = unchecked((int)0x80132001);
+        private const int CLR_E_BIND_ASSEMBLY_NOT_FOUND = unchecked((int)0x80132004);
+        private const int CLR_E_BIND_TYPE_NOT_FOUND = unchecked((int)0x80132005);
+
+        public static int BindAssembly(AssemblyBinder binder, AssemblyName assemblyName, bool excludeAppPaths, out Assembly? result)
+        {
+            int kContextVersion = 0;
+            BindResult bindResult = default;
+            int hr;
+            result = null;
+            ApplicationContext applicationContext = binder.AppContext;
+
+        // Tracing happens outside the binder lock to avoid calling into managed code within the lock
+        //BinderTracing::ResolutionAttemptedOperation tracer{ pAssemblyName, pBinder, 0 /*managedALC*/, hr};
```

Comment:
A question for how eventing works here:
It looks like that NativeRuntimeEventSource is the corresponding place for native events. It requires defining runtime-specific P/Invokes at NativeRuntimeEventSource.Threading.NativeSinks.CoreCLR.cs, and modifying the python script to avoid generating dummy methods. Is this the correct way?

Reply:
> Is this the correct way?

Yes, I think it is the correct way to do it.

If it works better, it should be fine to leave the fully managed binder tracing for a separate PR and call the unmanaged tracing using QCalls for now.